### PR TITLE
Unifying API for on-the-fly and gennodejs messages

### LINKED ISSUE #14: Unify API for on-the-fly generated and pre-compiled message classed
See https://github.com/RethinkRobotics-opensource/rosnodejs/pull/12#issuecomment-219899631.

This is complicated slightly by the fact that pre-compiled message classes can be instantiated synchronously, whereas on-the-fly generation of message classes is asynchronous. So a unified interface may need to follow the asynchronous approach, i.e., use a callback rather than a return value.

### PR DESCRIPTION
**Pull request by chfritz**
- returning promise from 'use'
- moved call to use into initNode

Unified APIs: rosnodejs.message now replaced with reosnodejs.require

Addresses part of #14.
- the way require now works is by first checking the registry of
  on-the-fly generated messages. If it finds the class there, it uses
  it, otherwise it will check the filesystem for the pre-compiled
  classes generated by gennodejs

### CODE REVIEW

**Comment by chfritz** [comment]
Fixed the conflict.

File: index.js [hunk 1]
```diff
@@ -181,17 +197,17 @@ let Rosnodejs = {
    * from .msg file if it doesn't already exist. This is mostly
    * because that would need to be async right now and we want a sync
    * method. */
-  message(type) {
-    return messages.getFromRegistry(type, "message");
-  },
+  // message(type) {
```
Comment (chris-smith) [inline index.js]: Any particular reason to keep this rather than pulling it?

File: index.js [hunk 2]
```diff
@@ -139,39 +150,44 @@ let Rosnodejs = {
...
+          return self._useServices(services);
+        }).then(() => {
+          resolve();
+        });
     });
   },
 
   /** create message classes for all the given types */
-  _useMessages(types, callback) {
+  _useMessages(types) {
```
Comment (chris-smith) [inline index.js:167]: can you update this and _useServices to handle the case where `types` is undefined? People may not provide those if they're using compiled messages.

**Comment by chris-smith** [comment]
This looks good! We should probably keep the bug open though because I believe there are still some differences to work out with respect to constructing messages (mostly with complex/nested types). I think some of that will require changes to `gennodejs` too.

**Comment by chfritz** [comment]
OK, done. Also removed the special "fixed #xx"  term to avoid auto-closing the issue.

File: index.js [hunk 3]
```diff
@@ -139,60 +150,54 @@ let Rosnodejs = {
...
-      console.log('get message type! ' + type);
-      messages.getMessage(type, function(error, Message) {
-        console.log('got it!');
-        Messages.push(Message);
-        if (Messages.length === types.length) {
-          callback();
-        }
+  _useMessages(types) {
+    if (!types || types.length == 0) {
+      resolve();
```
Comment (chris-smith) [inline index.js]: `return Promise.resolve()`

File: index.js [hunk 4]
```diff
@@ -139,60 +150,54 @@ let Rosnodejs = {
...
       });
     });
   },
-
+  
   /** create message classes for all the given types */
-  _useServices(types, callback) {
+  _useServices(types) {
+    if (!types || types.length == 0) {
+      resolve();
```
Comment (chris-smith) [inline index.js]: `return Promise.resolve()`

**Comment by chris-smith** [comment]
Tests and example run fine for me after `Promise.resolve` change - the service doesn't seem to be working in the on-the-fly example though. Could be an issue from the api update?

**Comment by chfritz** [comment]
Thanks for the fix. Yes, seems like I shouldn't be coding when tired. Also found the problem with the service -- another result of my sloppy merging. Should be good now.

**Comment by chris-smith** [comment]
No worries. Looks good now - thanks!

**Pull request merged**